### PR TITLE
using the structured semiring we can combine relation and relation ma…

### PR DESCRIPTION
**Pull request by swilly22**
…pping into a single matrix

### CODE REVIEW

File: src/graph/graph.c [hunk 1]
```diff
@@ -468,9 +444,10 @@ int Graph_GetEdgeRelation(const Graph *g, Edge *e) {
 
 	// Search for relation mapping matrix M, where
 	// M[dest,src] == edge ID.
-	for(int i = 0; i < array_len(g->_relations_map); i++) {
+	int relationship_count = array_len(g->relations);
+	for(int i = 0; i < relationship_count; i++) {
```
Comment (DvirDukhan) [inline src/graph/graph.c]: uint

File: src/graph/graph.c [hunk 2]
```diff
@@ -906,8 +873,8 @@ void _BulkDeleteEdges(Graph *g, Edge *edges, size_t edge_count) {
 		NodeID src_id = Edge_GetSrcNodeID(e);
 		NodeID dest_id = Edge_GetDestNodeID(e);
 		EdgeID edge_id;
-		GrB_Matrix M = Graph_GetRelationMap(g, r);  // Relation mapping matrix.
-		GrB_Matrix_extractElement_UINT64(&edge_id, M, src_id, dest_id);
+		GrB_Matrix R = Graph_GetRelationMatrix(g, r);  // Relation mapping matrix.
```
Comment (DvirDukhan) [inline src/graph/graph.c]: change comment

File: src/graph/graph.c [hunk 3]
```diff
@@ -967,28 +933,25 @@ void _BulkDeleteEdges(Graph *g, Edge *edges, size_t edge_count) {
...
 				// R = R & !mask.
-				GrB_Matrix_apply(R, mask, GrB_NULL, GrB_IDENTITY_BOOL, R, desc);
-				// M = M & !mask.
-				GrB_Matrix_apply(M, mask, GrB_NULL, GrB_IDENTITY_UINT64, M, desc);
+				GrB_Matrix_apply(R, mask, GrB_NULL, GrB_IDENTITY_UINT64, R, desc);
+
+				// Collect remaining edges. remaining_mask = remaining_mask + R.
+				GrB_eWiseAdd_Matrix_Semiring(remaining_mask, GrB_NULL, GrB_NULL, GxB_ANY_PAIR_BOOL, remaining_mask,
+											 mask, GrB_NULL);
+
```
Comment (DvirDukhan) [inline src/graph/graph.c]: This needs to be outside the `if(mask)`. You need to collect the remaining edges regardless to mask

File: src/graph/serializers/encoder/encode_graph.c [hunk 4]
```diff
@@ -188,12 +188,13 @@ void _RdbSaveEdges(RedisModuleIO *rdb, const Graph *g, char **string_mapping) {
 	// #edges (N)
 	RedisModule_SaveUnsigned(rdb, Graph_EdgeCount(g));
 
-	for(int r = 0; r < array_len(g->_relations_map); r++) {
+	int relationship_count = Graph_RelationTypeCount(g);
+	for(int r = 0; r < relationship_count; r++) {
```
Comment (DvirDukhan) [inline src/graph/serializers/encoder/encode_graph.c]: uint

File: src/graph/graph.c [hunk 5]
```diff
@@ -958,6 +923,7 @@ void _BulkDeleteEdges(Graph *g, Edge *edges, size_t edge_count) {
 	if(update_adj_matrices) {
 		GrB_Matrix remaining_mask;
 		GrB_Matrix_new(&remaining_mask, GrB_BOOL, Graph_RequiredMatrixDim(g), Graph_RequiredMatrixDim(g));
+
```
Comment (DvirDukhan) [inline src/graph/graph.c]: an empty line on purpose?

File: src/graph/graph.c (hunk 1 above)
Comment (swilly22) [reply src/graph/graph.c]: 👍

File: src/graph/graph.c (hunk 2 above)
Comment (swilly22) [reply src/graph/graph.c]: 👍

File: src/graph/graph.c (hunk 3 above)
Comment (swilly22) [reply src/graph/graph.c]: 👍

File: src/graph/serializers/encoder/encode_graph.c (hunk 4 above)
Comment (swilly22) [reply src/graph/serializers/encoder/encode_graph.c]: 👍

File: src/graph/graph.c (hunk 5 above)
Comment (swilly22) [reply src/graph/graph.c]: 👍

**Pull request merged**